examples/spark_udf/spark_udf_datetime.py

```diff
@@ -0,0 +1,65 @@
+import mlflow
+import pandas as pd
+from pyspark.sql import SparkSession
+from sklearn.datasets import load_iris
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.compose import ColumnTransformer
+from sklearn.preprocessing import FunctionTransformer
+from sklearn.pipeline import Pipeline
+
+
+def print_with_title(title, *args):
+    print(f"\n===== {title} =====\n")
+    for a in args:
+        print(a)
+
+
+def extract_month(df):
+    print_with_title("extract_month input", df.head(), df.dtypes)
+    transformed = df.assign(month=df["timestamp"].dt.month)
+    print_with_title("extract_month output", transformed.head(), transformed.dtypes)
+    return transformed
+
+
+def main():
+    X, y = load_iris(as_frame=True, return_X_y=True)
+    # Add a datetime column
+    months = X.index.to_series() % 12 + 1
+    timestamp = pd.to_datetime(months.map("2022-{:02d}-01 02:03:04".format))
+    X = X.assign(timestamp=timestamp)
+    print_with_title("Ran input", X.head(30), X.dtypes)
+
+    signature = mlflow.models.infer_signature(X, y)
+    print_with_title("Signature", signature)
+
+    month_extractor = FunctionTransformer(extract_month, validate=False)
+    timestamp_remover = ColumnTransformer(
+        [("selector", "passthrough", X.columns.drop("timestamp"))], remainder="drop"
+    )
```

Comment:
@harupy Rather than dropping the timestamp column, can we make sure that the model can predict on the timestamp properly? As we discussed offline, the model server should use the model schema to convert the timestamp column appropriately (i.e. if schema says the timestamp is a datetime, the column should be parsed as datetime; if the schema says the timestamp is a string, the column should be parsed as a string). Can we add those changes to this PR and test them?

Reply:
@dbczumar This is what `model.predict` receives on [this line](https://github.com/mlflow/mlflow/blob/86543a54405e7dd44df3fc3fc950e770c8e8aa70/mlflow/pyfunc/scoring_server/__init__.py#L308):

```
    sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)           timestamp
0                 5.1               3.5                1.4               0.2 2022-01-01 02:03:04
1                 4.9               3.0                1.4               0.2 2022-02-01 02:03:04
2                 4.7               3.2                1.3               0.2 2022-03-01 02:03:04
3                 4.6               3.1                1.5               0.2 2022-04-01 02:03:04
4                 5.0               3.6                1.4               0.2 2022-05-01 02:03:04
5                 5.4               3.9                1.7               0.4 2022-06-01 02:03:04
6                 4.6               3.4                1.4               0.3 2022-07-01 02:03:04
7                 5.0               3.4                1.5               0.2 2022-08-01 02:03:04
8                 4.4               2.9                1.4               0.2 2022-09-01 02:03:04
9                 4.9               3.1                1.5               0.1 2022-10-01 02:03:04
10                5.4               3.7                1.5               0.2 2022-11-01 02:03:04
11                4.8               3.4                1.6               0.2 2022-12-01 02:03:04
```

Note the timestamp column is dropped **in the model**. It must be removed because the KNeighborsClassifier


The type conversion is performed on this line:

https://github.com/mlflow/mlflow/blob/86543a54405e7dd44df3fc3fc950e770c8e8aa70/mlflow/utils/proto_json_utils.py#L224-L230

`dtypes` looks like this:

```
{'sepal length (cm)': dtype('float64'), 'sepal width (cm)': dtype('float64'), 'petal length (cm)': dtype('float64'), 'petal width (cm)': dtype('float64'), 'timestamp': dtype('<M8')}
```

Reply:
The `dtype('<M8')` shows it should read the timestamp data from JSON correctly.
We can add a unit test on `_dataframe_from_json` (assert the dataframe timestamp column values are correct)
instead of testing on model predicting on the timestamp properly.

Reply:
We'd better add similar unit test for all of datetime.datetime, datetime.date, datetime.time, pd.Timestamp.